Stop CWL from trying to copy downloaded directories from somewhere local

adamnovak opened the pull request:
[`cwltool` seems to treat `_:` as a special protocol for directories that need to be created rather than resolved to a filesystem path that gets copied.](https://github.com/common-workflow-language/cwltool/blob/9f3b9e7b74d5a904b12674dfd1300b56a48c3d33/cwltool/process.py#L311-L313) Since the Toil file store only stores files, and CWL ships around the whole metadata tree of files and directories needed to fulfill an InitialWorkDir input, we can just tell CWL that all the directories in the metadata come from the `_:` protocol.

For some reason, we can't just set all the directories' locations to just `_:`. They need to be unique for the `PathMapper` to be able to keep them straight. So I'm just prefixing `_:` onto the source `file:` URL that the directory had on the originating machine. This seems to be enough for CWL to both know that the directories are themselves and know that they don't need to be copied from anywhere.

I haven't tested this with any CWL workflows that actually drag along files in their directories (CI tests that, right?), but it works with the empty directory test from #2801. This should fix #2801.

mr-c commented on src/toil/cwl/cwltoil.py:
```diff
@@ -631,6 +646,40 @@ def write_file(writeFunc: Any, index: dict, existing: dict, file_uri: str) -> st
+    # doesn't have a real location. But each directory needs a unique location
+    # or CWL won't ship the metadata along. CWL takes "_:" as a signal to make
+    # directories instead of copying from somewhere. So we give every directory
+    # a unique _: location and CWL's machinery Just Works.
```
s/CWL/cwltool/g

mr-c commented on src/toil/cwl/cwltoil.py:
```diff
@@ -82,6 +82,8 @@
 from toil.common import Config, Toil, addOptions
 from toil.version import baseVersion
 
+logger = logging.getLogger(__name__)
```
Why not use `_logger` from https://github.com/DataBiosphere/toil/blob/dd7322725bac8fb65aacd70c3f3514982a2c8b2d/src/toil/cwl/cwltoil.py#L68 ?

adamnovak replied:
I set this up the way other Toil components do logging. I've never done logging the cwltool way before.

Presumably we should use only one of these in this module.

I actually would favor the Toil style as the right one to use here. The [cwltool logger setup](https://github.com/common-workflow-language/cwltool/blob/master/cwltool/loghandler.py) is pretty simple; it hardcodes INFO as the log level, logs everything as coming from "`cwltool`", and sends it to a default stream handler. The Toil style is to have loggers that report what file is doing the logging, and it obeys some options (`--logDebug`, `--logInfo`, etc.) to let the user control the log level.

Since Toil's `cwltoil.py` isn't actually `cwltool`, I think it makes sense for it to not log via `cwltool`'s internal logger. Moreover, since I'm adding some debug-level messages, I need a logger that I can set to debug level from the command line to actually see them, and the Toil logger setup has that, while I don't think cwltool's does.

Should Toil maybe be doing something to cwltool's `_logger` to adopt it and get its log messages to go through the same stream handler as Toil's?

adamnovak replied:
Yeah, if I send my debug-level logging to `cwllogger`, I don't see it even when running with `--logDebug`.

adamnovak commented:
OK @mr-c, I think I've addressed your comments.

I changed some references to "CWL" in general to now be about cwltool specifically.

I've also settled on pointing all the logging that `cwltoil.py` does itself to a Toil-style logger initialized with `__name__`, instead of creating log messages that claim to come from the `cwltool` module. But I've also fixed up the logging reconfiguration for the `cwltool` module so it happens in the worker processes as well as in the leader process. This makes Toil's `--logDebug` suddenly start showing the debug-level logging messages from `cwltool` in the individual job logs, like it should have been doing this whole time. It turns out `cwltool` logs a bunch of info about Files and Directories at DEBUG level already, so I cut back on some of my own added logging in favor of that.

Now it looks something like this:

```
DEBUG:cwltool:[job 3a931e67-06b1-4290-9e4d-4b6d9b4d585b] outputs {
    "ya": {
        "location": "file:///tmp/node-a9f628d3-fd18-44ac-aea0-8169c3407e32-11c2d78d2bcc48f28d74bd1f7dc0a45f/tmp6j7b_4tk/50a1e33b-5ccb-4bc2-bb1e-7f018716ad41/t2kvh9tuy/tmp-outuvoqo1t7/deeply/nested/dir/structure/ya",
        "basename": "ya",
        "nameroot": "ya",
        "nameext": "",
        "class": "File",
        "checksum": "sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709",
        "size": 0,
        "http://commonwl.org/cwltool#generation": 0
    }
}
DEBUG:cwltool:[job 3a931e67-06b1-4290-9e4d-4b6d9b4d585b] Removing input staging directory /tmp/tmpvb4k026e
DEBUG:toil.cwl.cwltoil:Sending file at: toilfs:0:files/for-job/kind-file_z_home_anovak_workspace_toil_test_v1.0_v1.0_iwdr_with_nested_dirs.cwl_second_3a931e67-06b1-4290-9e4d-4b6d9b4d585b/instance-395aiuba/file-nbpf8szh/ya
DEBUG:toil.fileStores.abstractFileStore:LOG-TO-MASTER: Job files/for-job/kind-CWLWorkflow/instance-2ssncyrt/cleanup/file-epir9qk1/stream used 0.00% (52.0 KB [53248B] used, 3.0 GB [3221225472B] requested) at the end of its run.
```
